File: src/com/train/algorithm/search/implementInC++/PermutationsII.cpp
//
// Created by 罗斯 on 11/9/2022.
//
#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        sort(begin(nums), end(nums));
        vector<vector<int>> res;
        vector<int> used(nums.size());
        vector<int> cur;
        dfs(nums,cur,res,used);
        return res;
    }

private:
    void dfs(const vector<int>& nums,vector<int>& cur,vector<vector<int>>& res,vector<int>& used){
        if (cur.size() == nums.size()){
            res.push_back(cur);
            return;
        }

        for (int i = 0; i < nums.size(); ++i) {
            if(used[i]) continue;
            if(i > 0 && nums[i] == nums[i - 1] && !used[i-1]) continue;
            used[i] = 1;
            cur.push_back(nums[i]);
            dfs(nums,cur,res,used);
            cur.pop_back();
            used[i] = 0;
        }
    }
};

File: src/com/train/algorithm/search/implementInPython/PermutationsII.py
class Solution(object):
    def permuteUnique(self, nums):
        ans = []
        n = len(nums)
        visited = [False] * len(nums)

        nums.sort()

        def backtracking(nums, res):

            if len(res) == n:
                ans.append(res[:])
                return

            for i in range(len(nums)):

                if visited[i] or (i > 0 and nums[i] == nums[i - 1] and not visited[i - 1]):
                    continue

                res.append(nums[i])

                visited[i] = True
                backtracking(nums, res)

                visited[i] = False
                res.pop()

        backtracking(nums, [])
        return ans